File: python_base_04/core/recall_game/models/card.py
"""
Card Models for Recall Game

This module defines the card system for the Recall card game,
including standard cards, special power cards, and point calculations.
"""

from typing import Optional, List, Dict, Any
from enum import Enum
import random


class CardSuit(Enum):
    """Card suits"""
    HEARTS = "hearts"
    DIAMONDS = "diamonds"
    CLUBS = "clubs"
    SPADES = "spades"


class CardRank(Enum):
    """Card ranks with their point values"""
    JOKER = "joker"
    ACE = "ace"
    TWO = "2"
    THREE = "3"
    FOUR = "4"
    FIVE = "5"
    SIX = "6"
    SEVEN = "7"
    EIGHT = "8"
    NINE = "9"
    TEN = "10"
    JACK = "jack"
    QUEEN = "queen"
    KING = "king"


class SpecialPowerType(Enum):
    """Special power card types"""
    PEEK_AT_CARD = "peek_at_card"  # Queen
    SWITCH_CARDS = "switch_cards"   # Jack
    STEAL_CARD = "steal_card"       # Added power card
    DRAW_EXTRA = "draw_extra"       # Added power card
    PROTECT_CARD = "protect_card"   # Added power card


class Card:
    """Represents a single card in the Recall game"""
    
    def __init__(self, rank: str, suit: str, points: int, 
                 special_power: Optional[str] = None, card_id: Optional[str] = None):
        self.rank = rank
        self.suit = suit
        self.points = points
        self.special_power = special_power
        self.card_id = card_id or f"{rank}_{suit}_{random.randint(1000, 9999)}"
        self.is_visible = False  # Whether the card is face up
        self.owner_id = None     # Player who owns this card
    
    def __str__(self):
        if self.rank == "joker":
            return "Joker"
        return f"{self.rank.title()} of {self.suit.title()}"
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert card to dictionary representation"""
        return {
            "card_id": self.card_id,
            "rank": self.rank,
            "suit": self.suit,
            "points": self.points,
            "special_power": self.special_power,
            "is_visible": self.is_visible,
            "owner_id": self.owner_id
        }
    
    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'Card':
        """Create card from dictionary"""
        card = cls(
            rank=data["rank"],
            suit=data["suit"],
            points=data["points"],
            special_power=data.get("special_power"),
            card_id=data.get("card_id")
        )
        card.is_visible = data.get("is_visible", False)
        card.owner_id = data.get("owner_id")
        return card
    
    def get_point_value(self) -> int:
        """Get the point value of this card"""
        return self.points
    
    def has_special_power(self) -> bool:
        """Check if this card has a special power"""
        return self.special_power is not None
    
    def can_play_out_of_turn(self, played_card: 'Card') -> bool:
        """Check if this card can be played out of turn"""
        return self.rank == played_card.rank


class CardDeck:
    """Represents a deck of cards for the Recall game"""
    
    def __init__(self, include_jokers: bool = True, include_special_powers: bool = True):
        self.cards = []
        self.include_jokers = include_jokers
        self.include_special_powers = include_special_powers
        self._initialize_deck()
    
    def _initialize_deck(self):
        """Initialize the deck with all cards"""
        # Standard 52-card deck
        suits = ["hearts", "diamonds", "clubs", "spades"]
        ranks = ["ace", "2", "3", "4", "5", "6", "7", "8", "9", "10", "jack", "queen", "king"]
        
        # Add standard cards
        for suit in suits:
            for rank in ranks:
                points = self._get_point_value(rank, suit)
                special_power = self._get_special_power(rank)
                card = Card(rank, suit, points, special_power)
                self.cards.append(card)
        
        # Add Jokers (0 points)
        if self.include_jokers:
            for i in range(2):  # 2 jokers
                card = Card("joker", "joker", 0, None)
                self.cards.append(card)
        
        # Add special power cards (if enabled)
        if self.include_special_powers:
            self._add_special_power_cards()
    
    def _get_point_value(self, rank: str, suit: str) -> int:
        """Get the point value for a card"""
        if rank == "joker":
            return 0
        elif rank == "ace":
            return 1
        elif rank in ["2", "3", "4", "5", "6", "7", "8", "9", "10"]:
            return int(rank)
        elif rank in ["jack", "queen"]:
            return 10
        elif rank == "king":
            # Red kings (hearts, diamonds) are 0 points
            if suit in ["hearts", "diamonds"]:
                return 0
            else:
                return 10
        else:
            return 0
    
    def _get_special_power(self, rank: str) -> Optional[str]:
        """Get special power for a card"""
        if rank == "queen":
            return "peek_at_card"
        elif rank == "jack":
            return "switch_cards"
        else:
            return None
    
    def _add_special_power_cards(self):
        """Add special power cards to the deck"""
        special_powers = [
            ("steal_card", 5),
            ("draw_extra", 3),
            ("protect_card", 4),
            ("skip_turn", 6),
            ("double_points", 8)
        ]
        
        for power, points in special_powers:
            # Create special power cards with unique suits
            for i in range(2):  # 2 of each special power
                card = Card(f"power_{power}", f"special_{i}", points, power)
                self.cards.append(card)
    
    def shuffle(self):
        """Shuffle the deck"""
        random.shuffle(self.cards)
    
    def draw_card(self) -> Optional[Card]:
        """Draw a card from the top of the deck"""
        if self.cards:
            return self.cards.pop()
        return None
    
    def add_card(self, card: Card):
        """Add a card to the deck"""
        self.cards.append(card)
    
    def get_remaining_count(self) -> int:
        """Get the number of cards remaining in the deck"""
        return len(self.cards)
    
    def is_empty(self) -> bool:
        """Check if the deck is empty"""
        return len(self.cards) == 0
    
    def reset(self):
        """Reset the deck to its original state"""
        self.cards = []
        self._initialize_deck() 

File: flutter_base_05/lib/core/recall_game/models/card.dart
import 'package:flutter/material.dart' hide Card;
import '../utils/game_constants.dart';

/// Card suit enumeration
enum CardSuit {
  hearts,
  diamonds,
  clubs,
  spades,
}

/// Card rank enumeration
enum CardRank {
  ace(1),
  two(2),
  three(3),
  four(4),
  five(5),
  six(6),
  seven(7),
  eight(8),
  nine(9),
  ten(10),
  jack(11),
  queen(12),
  king(13);

  const CardRank(this.value);
  final int value;
}

/// Special power types for cards
enum SpecialPowerType {
  queen,    // Peek at a card
  jack,     // Switch cards
  addedPower, // Custom special powers
  none,     // No special power
}

/// Card model for the Recall game
class Card {
  final CardSuit suit;
  final CardRank rank;
  final int points;
  final SpecialPowerType specialPower;
  final String? specialPowerDescription;
  final Map<String, dynamic>? specialPowerData;
  
  const Card({
    required this.suit,
    required this.rank,
    required this.points,
    this.specialPower = SpecialPowerType.none,
    this.specialPowerDescription,
    this.specialPowerData,
  });

  /// Get card display name
  String get displayName {
    final rankName = rank.name.substring(0, 1).toUpperCase() + rank.name.substring(1);
    final suitSymbol = _getSuitSymbol();
    return '$rankName$suitSymbol';
  }

  /// Get suit symbol for display
  String _getSuitSymbol() {
    switch (suit) {
      case CardSuit.hearts:
        return '♥';
      case CardSuit.diamonds:
        return '♦';
      case CardSuit.clubs:
        return '♣';
      case CardSuit.spades:
        return '♠';
    }
  }

  /// Get card color (red or black)
  String get color {
    return (suit == CardSuit.hearts || suit == CardSuit.diamonds) ? 'red' : 'black';
  }

  /// Check if card has special power
  bool get hasSpecialPower => specialPower != SpecialPowerType.none;

  /// Check if card can be played out of turn
  bool get canPlayOutOfTurn {
    return hasSpecialPower && 
           (specialPower == SpecialPowerType.queen || 
            specialPower == SpecialPowerType.jack ||
            specialPower == SpecialPowerType.addedPower);
  }

  /// Convert card to JSON
  Map<String, dynamic> toJson() {
    return {
      'suit': suit.name,
      'rank': rank.name,
      'points': points,
      'specialPower': specialPower.name,
      'specialPowerDescription': specialPowerDescription,
      'specialPowerData': specialPowerData,
      'displayName': displayName,
      'color': color,
    };
  }

  /// Create card from JSON
  factory Card.fromJson(Map<String, dynamic> json) {
    return Card(
      suit: CardSuit.values.firstWhere((s) => s.name == json['suit']),
      rank: CardRank.values.firstWhere((r) => r.name == json['rank']),
      points: json['points'],
      specialPower: SpecialPowerType.values.firstWhere((s) => s.name == json['specialPower']),
      specialPowerDescription: json['specialPowerDescription'],
      specialPowerData: json['specialPowerData'],
    );
  }

  /// Create a standard deck of 52 cards
  static List<Card> createStandardDeck() {
    final List<Card> deck = [];
    
    for (final suit in CardSuit.values) {
      for (final rank in CardRank.values) {
        int points = _calculatePoints(rank);
        SpecialPowerType specialPower = _getSpecialPower(rank);
        String? specialPowerDescription = _getSpecialPowerDescription(rank);
        
        deck.add(Card(
          suit: suit,
          rank: rank,
          points: points,
          specialPower: specialPower,
          specialPowerDescription: specialPowerDescription,
        ));
      }
    }
    
    return deck;
  }

  /// Calculate points for a card rank
  static int _calculatePoints(CardRank rank) {
    switch (rank) {
      case CardRank.ace:
        return 1;
      case CardRank.two:
      case CardRank.three:
      case CardRank.four:
      case CardRank.five:
      case CardRank.six:
      case CardRank.seven:
      case CardRank.eight:
      case CardRank.nine:
      case CardRank.ten:
        return rank.value;
      case CardRank.jack:
        return 11;
      case CardRank.queen:
        return 12;
      case CardRank.king:
        return 13;
    }
  }

  /// Get special power for a card rank
  static SpecialPowerType _getSpecialPower(CardRank rank) {
    switch (rank) {
      case CardRank.queen:
        return SpecialPowerType.queen;
      case CardRank.jack:
        return SpecialPowerType.jack;
      default:
        return SpecialPowerType.none;
    }
  }

  /// Get special power description
  static String? _getSpecialPowerDescription(CardRank rank) {
    switch (rank) {
      case CardRank.queen:
        return 'Peek at a card from any player\'s hand';
      case CardRank.jack:
        return 'Switch a card with another player';
      default:
        return null;
    }
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is Card &&
        other.suit == suit &&
        other.rank == rank;
  }

  @override
  int get hashCode => suit.hashCode ^ rank.hashCode;

  @override
  String toString() {
    return 'Card($displayName, points: $points)';
  }
}

/// Card deck model for managing a collection of cards
class CardDeck {
  List<Card> _cards;

  CardDeck({List<Card>? cards}) : _cards = cards ?? [];

  /// Get all cards in the deck
  List<Card> get cards => List.unmodifiable(_cards);

  /// Get number of cards in deck
  int get length => _cards.length;

  /// Check if deck is empty
  bool get isEmpty => _cards.isEmpty;

  /// Check if deck is not empty
  bool get isNotEmpty => _cards.isNotEmpty;

  /// Shuffle the deck
  void shuffle() {
    _cards.shuffle();
  }

  /// Draw a card from the top of the deck
  Card? drawCard() {
    if (_cards.isEmpty) return null;
    return _cards.removeAt(0);
  }

  /// Draw multiple cards
  List<Card> drawCards(int count) {
    final drawnCards = <Card>[];
    for (int i = 0; i < count && _cards.isNotEmpty; i++) {
      drawnCards.add(_cards.removeAt(0));
    }
    return drawnCards;
  }

  /// Add a card to the bottom of the deck
  void addCard(Card card) {
    _cards.add(card);
  }

  /// Add multiple cards to the deck
  void addCards(List<Card> cards) {
    _cards.addAll(cards);
  }

  /// Clear all cards from the deck
  void clear() {
    _cards.clear();
  }

  /// Create a standard deck and shuffle it
  factory CardDeck.standard() {
    final deck = CardDeck(cards: Card.createStandardDeck());
    deck.shuffle();
    return deck;
  }

  /// Convert deck to JSON
  Map<String, dynamic> toJson() {
    return {
      'cards': _cards.map((card) => card.toJson()).toList(),
    };
  }

  /// Create deck from JSON
  factory CardDeck.fromJson(Map<String, dynamic> json) {
    final cards = (json['cards'] as List)
        .map((cardJson) => Card.fromJson(cardJson))
        .toList();
    return CardDeck(cards: cards);
  }
} 